RFC-0022-Model-Partitioning-in-Pipeline-Parallelism.md

```diff
@@ -0,0 +1,255 @@
+
+We will use torch.fx to do a profiling run to collect the statistics (i.e. execution_time, parameter_size, execution_order, etc.). The profiler still needs user to pass in a sample input, and we do a full run on the model base on this sample input. We can adjust the statistics to collect base on the partition algorithm we choose (i.e. module-level statistics or op-level statistics)
+
+```
+from torch.fx import Interpreter
+
+class ProfilingInterpreter(Interpreter):
+
+    def __init__(self, mod : torch.nn.Module):
+        gm = torch.fx.symbolic_trace(mod)
+        super().__init__(gm)
+
+        # We are going to store away three things here:
+        #
+        # 1. execution time of each module/node
+        # 2. parameter_sizes of each module/node
+        # 3. activation_sizes of each module
+        self.execution_times_sec : Dict[str, float] = []
+        self.parameter_sizes : Dict[str, float] = {}
+        self.activation_sizes: Dict[str, float] = {}
+
+    def run_node(self, n : torch.fx.Node) -> Any:
+        # Record the time we started running the op
+        t_start = time.time()
+        # Run the op
+        return_val = super().run_node(n)
+        # Record the time we finished running the op
+        t_end = time.time()
+        self.execution_times_sec[module_qual_name] = t_end - t_start
+
+        # also update the parameter_size and activation_size for
+        # module nodes 
+
+        return return_val
+```
+
+
+
+
+
+## Partition Algorithm
+
+There’re several partition algorithms as we mentioned in the relevant works section, each of them have its own pros/cons. For example, torchgpipe/fairscale only allows partitioning the model on the top level, SageMaker only partition on the module level instead of operation level. So there might be some unbalanced cases for those approaches.  DAPPLE explores a bigger search space by controlling the stages/schedules, but the algorithm complexity is too high. 
+
+Our approach here starts with an approach similar to fairscale, which only do top-level partition, but it could be improved further by exploring more partition approaches, i.e. with operation level partitioning by torch.fx, we could partition the model in a more even manner, which further improves the efficiency. Since we only expose a simple API and return the partitioned model with stages assigned to the devices, the internal implementation can be improved as we explore more ideas. The plan for partition algorithm exploration:
+
+* Use fairscale/gpipe's [Block Partition of Sequences](https://arxiv.org/pdf/1308.2452.pdf) as our first step
+* See if we can apply [Fiduccia-Mattheyses Heuristic](https://en.wikipedia.org/wiki/Fiduccia%E2%80%93Mattheyses_algorithm) to partition the graph and how the performance compare with the default one
+* See if we can do operation-level tracing, and partition the model into several balanced `torch.fx.GraphModule` instead of using the original module architecture
```

Comment:
Would that mean that we lose all methods linked to the original class? For example `transformers` fit a certain interface which allows them to use a `generate` method. If one uses `torch.fx.GraphModule` it loses all methods linked to it's original module?